### DISCUSSION TITLE
SQL UPDATE

### DISCUSSION THREAD

**Discussion by ondohotola** [dathere/qsv#1270] [Q&A] [answered]
Given two files which *should* but are not guaranteed to be of equal lengths as they are generated after another (DomainsInclude.csv and DomainsExclude.csv) I need to set a column `eppstatus` in the one to the value of same in the other.

The following CSVQ command works but is of course extremely slow with 20000 lines (with or without the ORDER):

```sql
UPDATE DomainsInclude 
SET eppstatus = (
   SELECT DomainsExclude.eppstatus 
   FROM DomainsExclude 
   WHERE DomainsInclude.name = DomainsExclude.name
   ORDER BY DomainsExclude.name
);
COMMIT;
```

Is there a `qsv` pipeline to do this (so I can test for speed)?

`sqlp` does not seem to have such an `update` command.

**Comment by jqnatividad** [comment, marked as answer]
Hi @ondohotola , I can think of two solutions:

1. `sqlp` with `joinp`:

First, do the SELECT with sqlp:
```bash
qsv sqlp DomainExclude.csv DomainsInclude.csv \
  "select DomainsExclude.name, DomainsExclude.appstatus \
   FROM DomainsExclude, DomainsInclude \
   WHERE DomainsInclude.name = DomainsExclude.name \
   ORDER BY DomainsExclude.name" \
  --output DE_appstatus.csv
```
and then `joinp` it with DomainsInclude
```bash
qsv joinp name DomainsInclude.csv name DE_appstatus.csv --output joined_with_appstatus.csv
```

2. create a SQL script and do it with one `sqlp` operation..

  Create a SQL script named Domains.sql:
  ```SQL
-- first, create a table from the select
CREATE TABLE DomainsExclude_AppStatus as 
  SELECT DomainsExclude.name, DomainsExclude.appstatus
   FROM DomainsExclude, DomainsInclude
   WHERE DomainsInclude.name = DomainsExclude.name
   ORDER BY DomainsExclude.name;

-- then we do the join with the just created table
SELECT *
  FROM DomainsInclude JOIN DomainsExclude_AppStatus
  ON DomainsInclude.name = DomainsExclude_AppStatus.name;
  ```
  
and then run it with sqlp:

  ```bash
qsv sqlp DomainsExclude.csv DomainsInclude.csv Domains.sql --output Domains_Appstatus.csv
  ```

Let me know if either of this works and how fast it is compared with CSVQ.

**Comment by ondohotola** [comment]
option 1 takes seconds :-)-O (`csvq` takes 3 minutes) thank you very much

**Reply by jqnatividad** [reply]
Great to know!  What about option 2?  It should also work.  Also, can you share the specific timings?  I'm working on publishing benchmarks again as part of the release process and am keen to learn how qsv performs compared with other tools.

**Reply by ondohotola** [reply]
The author of CSVQ [@mithrandie](https://github.com/mithrandie) let me very gracefully have some more SQL

```sql
UPDATE i
   SET i.eppstatus = e.eppstatus
  FROM DomainsInclude i
       JOIN DomainsExclude e
         ON e.name = i.name
```

stating it might not be fast enough, which reduced from 580 seconds to 132 and would be sufficient for my 19247 lines.

Your option 1 takes 0.5 seconds which is astonishing :-)-O 

I am noticing some issues with your second option, doesn't seem to work at all, but don't have any more time looking at this.

All of this running on a 2020 13" Intel MacBook Pro with 32GB RAM

Thank you

**Reply by jqnatividad** [reply]
Wow! That's blazing-fast!!! 🚀 And it's really more a testament to the power of Polars' LazyFrame expressions.

Anyways, Solution 2 should be at least as fast, if not faster because you don't have to call qsv twice.

Here's a simplified SQL script to try which I know works because I tried it on dummy data.

Create a file named DI_status.sql
```SQL
CREATE TABLE DE_AppStatus as 
  SELECT DE.name, DE.appstatus 
   FROM DE, DI
   WHERE DI.name = DE.name
   ORDER BY DE.name;

SELECT *
  FROM DI join DE_AppStatus
  ON DI.name = DE_AppStatus.name;
```

And run it using:
```bash
$ qsv sqlp DI.csv DE.csv DI_status.sql
```

**Reply by derekmahar** [reply]
How did you generate your dummy data?

**Reply by ondohotola** [reply]
The _simplified_ script is actually Option 2 which is the fastest (ie 0.3 seconds for 19000 lines).

I am really impressed.

**Reply by derekmahar** [reply]
How long does it take the simplified script to run in DuckDB?

**Reply by jqnatividad** [reply]
Hi @derekmahar ,

The dummy data was just a few lines:

[DE.csv](https://github.com/jqnatividad/qsv/files/12508366/DE.csv)
[DI.csv](https://github.com/jqnatividad/qsv/files/12508367/DI.csv)

Basically, just having enough data to retrieve appstatus from DomainsExclude.appstatus to join with DomainsInclude.

**Comment by ondohotola** [comment]
Ok,

will measure later in the PM.

**Comment by derekmahar** [comment]
Can you share any sample data files?

**Reply by ondohotola** [reply]
No, NDA

**Reply by derekmahar** [reply]
What about column headers with obfuscated, fake, or random data?

**Reply by derekmahar** [reply]
Have you tried DuckDB just to see how its performance compares to that of csvq and `qsv sqlp`?

**Comment by ondohotola** [comment]
Too much work, almost 100 columns and almost all of them GDPR'ed.

**Reply by jqnatividad** [reply]
One more reason I should work on the `fake` command - https://github.com/jqnatividad/qsv/issues/235 😉

`fake` scans your data, compile stats and frequency tables on it, and use that metadata to generate fake data that has the same schema and similar summary statistics - but using [faker](https://github.com/cksac/fake-rs#fakers-with-locale) keywords for certain field classes to ensure privacy.

**Reply by ondohotola** [reply]
Great idea

**Comment by ondohotola** [comment]
Can you propose some command?

**Reply by derekmahar** [reply]
For a performance comparison, I propose importing your CSV files into DuckDB and then executing the [SQL update command](https://github.com/mithrandie/csvq/discussions/116#discussioncomment-6892093) that Yuki suggested.

**Reply by ondohotola** [reply]
Takes about 20 seconds

**Reply by derekmahar** [reply]
How does this compare to csvq and qsv?

**Reply by derekmahar** [reply]
Nevermind. I just read the times that you reported above.

**Reply by derekmahar** [reply]
> For a performance comparison, I propose importing your CSV files into DuckDB and then executing the [SQL update command](https://github.com/mithrandie/csvq/discussions/116#discussioncomment-6892093) that Yuki suggested.

Performance tests using DuckDB v0.8.2-dev3949 in an Ubuntu 22.04 virtual machine running on an AMD Ryzen 9 6900HX with a Samsung SSD 980 PRO 1TB NVMe drive::

```
$ ./create_csv_files
$ time ./run_test_1

real    0m0.139s
user    0m0.055s
sys     0m0.020s
$ time ./run_test_2

real    0m0.143s
user    0m0.050s
sys     0m0.026s
```
Test times include importing the CSV files into tables in a DuckDB database.

See scripts in [duckdb_update_domains.zip](https://github.com/jqnatividad/qsv/files/12516089/duckdb_update_domains.zip).

**Reply by derekmahar** [reply]
Performance tests using csvq 1.17.11 in an Ubuntu 22.04 virtual machine running on an AMD Ryzen 9 6900HX with a Samsung SSD 980 PRO 1TB NVMe drive: 

```
$ ./create_csv_files
$ time ./run_test_1
20000 records updated on "/home/derek/Data/Test/csvq/DomainsInclude.csv".
Rollback: file "/home/derek/Data/Test/csvq/DomainsInclude.csv" is restored.

real    0m58.689s
user    3m45.655s
sys     0m2.454s
$ time ./run_test_2
20000 records updated on "/home/derek/Data/Test/csvq/DomainsInclude.csv".
Rollback: file "/home/derek/Data/Test/csvq/DomainsInclude.csv" is restored.

real    0m42.064s
user    3m21.352s
sys     0m1.205s
```

Note that each test rolls back the changes to `DomainsInclude.csv` so that each subsequent test may reuse the original version of this data file.

See scripts in [csvq_update_domains.zip](https://github.com/jqnatividad/qsv/files/12516129/csvq_update_domains.zip).

**Reply by jqnatividad** [reply]
Thanks @derekmahar for preparing this.

It's still good to know that `qsv sqlp` is holding its own compared to these tools with a minimum of preparation - right from the command-line with one command, without importing the data first into a database.

**Reply by derekmahar** [reply]
Performance test using qsv 0.112.0 in an Ubuntu 22.04 virtual machine running on an AMD Ryzen 9 6900HX with a Samsung SSD 980 PRO 1TB NVMe drive:

```
$ time ./run_test_1
(20000, 2)
(20000, 3)

real    0m0.052s
user    0m0.044s
sys     0m0.030s
```

See scripts in [qsv_update_domains.zip](https://github.com/jqnatividad/qsv/files/12516237/qsv_update_domains.zip).

**Reply by derekmahar** [reply]
Test times using DuckDB, but excluding the time to import the CSV files into DuckDB tables:

```
$ time ./run_test_1

real    0m0.047s
user    0m0.002s
sys     0m0.016s
$ time ./run_test_2

real    0m0.049s
user    0m0.012s
sys     0m0.007s
```

**Reply by jqnatividad** [reply]
Can you humor me @derekmahar and modify the `run_test_1` script for qsv to run without printing to stdout with the `--quiet` option?

It's already plenty fast and was just wondering what fraction of time it spends pushing out the shape of the result to stdout.

**Reply by derekmahar** [reply]
qsv test time using the `--quiet` option:

```
$ time ./run_test_1

real    0m0.051s
user    0m0.056s
sys     0m0.017s
```

**Reply by derekmahar** [reply]
Note that all test times include the time to execute bash.

**Reply by jqnatividad** [reply]
Great stuff!

You may want to look into using [hyperfine](https://github.com/sharkdp/hyperfine#hyperfine) - a CLI tool purpose made for benchmarking.

Its statistical analysis should help tease out the times introduced by running bash, especially at these sub second response times.

**Reply by derekmahar** [reply]
Thank you for the tip about hyperfine.  Given that each test script invokes bash, how does hyperfine exclude the start-up time that bash imposes?

**Reply by jqnatividad** [reply]
Because it can do warm-ups, do multiple runs, and clear the cache - it should help establish a common/fair baseline for any "noise" introduced by bash/system IO/operating system stuff.
It will also allow you to compare multiple commands side-by-side.

**Reply by derekmahar** [reply]
Are you referring to the [intermediate shell](https://github.com/sharkdp/hyperfine#intermediate-shell) that hyperfine invokes?  I'm referring to the bash shell that each of my test scripts invoke since most of the scripts invoke multiple commands.

**Reply by derekmahar** [reply]
Alternatively, I could define my tests as [shell functions](https://github.com/sharkdp/hyperfine#shell-functions-and-aliases) and export these into the current shell where hyperfine can find them.

**Comment by ondohotola** [comment]
didn't try

**Comment by derekmahar** [comment]
> It's still good to know that `qsv sqlp` is holding its own compared to these tools with a minimum of preparation - right from the command-line with one command, without importing the data first into a database.

Two commands if you count `qsv sqlp` and `qsv joinp`.

**Reply by jqnatividad** [reply]
Yes @derekmahar ,
Solution 1 was two commands.

Solution 2, however, is just one `sqlp` command invoking a SQL script with two SQL statements.

**Reply by derekmahar** [reply]
My qsv test invoked only solution 1.  I'll create a second script that invokes solution 2.

**Reply by derekmahar** [reply]
```
$ hyperfine ./run_test_1 ./run_test_2
Benchmark 1: ./run_test_1
  Time (mean ± σ):      50.8 ms ±   0.6 ms    [User: 47.2 ms, System: 22.7 ms]
  Range (min … max):    49.6 ms …  52.8 ms    57 runs

Benchmark 2: ./run_test_2
  Time (mean ± σ):      26.8 ms ±   0.7 ms    [User: 26.1 ms, System: 12.1 ms]
  Range (min … max):    25.5 ms …  31.2 ms    110 runs

Summary
  './run_test_2' ran
    1.89 ± 0.05 times faster than './run_test_1'
```

See scripts in [qsv_update_domains.zip](https://github.com/jqnatividad/qsv/files/12516592/qsv_update_domains.zip).

**Reply by jqnatividad** [reply]
Nice! Soln 2 ran almost twice as fast than Soln 1.

That tells me its more of the preprocessing done by the OS/file IO/qsv, since at the heart of it - both solutions do exactly the same thing - 1) do a Select into a temporary table, and then 2) join the temporary table with one of the input files.

So if say the startup/preprocessing time is ~10 ms:
* for Soln 1 - it takes 50 ms, with two commands. ~20 ms for startup/preprocessing `joinp`, then `sqlp`, - that leaves about 30ms for processing.
* for Soln 2- 25 ms. ~10 ms for startup/preprocessing, that leaves 10-15 ms.

10-15 milliseconds to do the actual "work". 😄 

Of course, these are all back-of-the-napkin guesstimates - we can actually capture the elapsed time in the logs by setting the environment variable `QSV_LOG_LEVEL=info`.

**Reply by derekmahar** [reply]
I noticed that neither [Solution 1](https://github.com/jqnatividad/qsv/discussions/1270#discussioncomment-6893060) nor [Solution 2](https://github.com/jqnatividad/qsv/discussions/1270#discussioncomment-6897817) produces the same result as the [original query](https://github.com/jqnatividad/qsv/discussions/1270#discussion-5584462) because they exclude from their results rows that are present in `DomainsInclude.csv`, but not in `DomainsExclude.csv`.  In order to include these rows, I changed the `JOIN` operations in these solutions to `LEFT OUTER JOIN` operations:

Solution 1:

```bash
qsv sqlp DomainsExclude.csv DomainsInclude.csv \
  "SELECT DomainsExclude.name, DomainsExclude.eppstatus \
   FROM DomainsExclude, DomainsInclude \
   WHERE DomainsInclude.name = DomainsExclude.name \
   ORDER BY DomainsExclude.name" \
  --output DomainsExclude_appstatus.csv \
  --quiet

qsv joinp --left --quiet \
  name DomainsInclude.csv \
  name DomainsExclude_appstatus.csv
```
Solution 2:

```sql
CREATE TABLE DE_AppStatus as 
SELECT DE.name, DE.appstatus
FROM DE, DI
WHERE DI.name = DE.name
ORDER BY DE.name;

SELECT *
FROM DI
  LEFT OUTER JOIN DE_AppStatus
  ON DI.name = DE_AppStatus.name;
```

See revised input data files and test scripts in [qsv_update_domains.zip](https://github.com/jqnatividad/qsv/files/12527704/qsv_update_domains.zip).

**Reply by derekmahar** [reply]
By the way @jqnatividad, when using `qsv joinp`, how do I select the particular columns that the join operation will display in the result?

**Reply by jqnatividad** [reply]
Currently, you can't. It just returns all the columns. Polars is just smart enough to add the "_left", "_right" suffix if there are duplicate column names.

I was thinking folks can just use `qsv select` afterwards to remove the unwanted columns.

But yeah, I'll be sure to add output column selection to the backlog

**Reply by jqnatividad** [reply]
Actually @derekmahar , there is no specific command-line option to select the result columns.

It just returns the columns based on the LAST query in the SQL script we used.  And since we did "SELECT *", it returned all the columns of the JOIN query, with the "_right" suffix to differentiate duplicate columns.

You can limit the columns by changing the SELECT.  You can even do "SELECT colname AS othername" to rename columns.

**Reply by derekmahar** [reply]
Are you referring to `qsv sqlp`?  I was referring to `qsv joinp`.

**Reply by jqnatividad** [reply]
My mistake... I was thinking `sqlp` - `joinp` effectively does `SELECT *` by default...

which gives me an idea how to enable column selection for `joinp` - filter the result using a SQLcontext.... 😉

**Reply by jqnatividad** [reply]
And `joinp` now has the `--sql-filter` option.

https://github.com/jqnatividad/qsv/pull/1287

**Comment by ondohotola** [comment]
That _right is rather helpful, actually.

**Comment by ondohotola** [comment]
It might be interesting to know that since `sqlp` doesn't do subqueries I have used CSVQ with something like this 

```sql
SELECT
   DISTINCT clients.Name as `COMPANY NAME`,
   clients.`Client ID` as `Client Identifier`,
   clients.`E-mail` as `Main E-mail`,
   "no" as `Connect via EPP`,
   clients.Country as Country,
   clients.Phone as Phone,
   clients.`Admin Contact` as `Administrative Contact`,
   clients.`E-mail` as `Administrative E-mail`
FROM
   Clients clients
WHERE
   EXISTS ( SELECT 1
            FROM DomainsInclude domains
            WHERE clients.`Client ID` = domains.`registrar_id` )
ORDER BY
   clients.Name,
   clients.`Client ID`
```

which on a small `clients.csv` and the 19000 `DomainsInclude.csv` file takes 5 seconds.

Last night I was bored and figured out `joinp` and am pleased to report that this takes 200 milliseconds

**Reply by derekmahar** [reply]
What `joinp` command did you use?  `sqlp` can perform join operations, too.

**Reply by jqnatividad** [reply]
Actually, `sqlp` supports [Common Table Expressions](https://learnsql.com/blog/sql-subquery-cte-difference/) @ondohotola ...

```SQL
WITH millennials AS (SELECT * FROM people WHERE age >= 25 and age <= 40) \
    SELECT * FROM millennials WHERE owns_home = 'N'
```

```bash
qsv sqlp people.csv "WITH millennials AS (SELECT * FROM people WHERE age >= 25 and age <= 40) \
    SELECT * FROM millennials WHERE owns_home = 'N'"
```

Myself, I actually prefer using SQL scripts with multiple SQL statements with `sqlp` instead of CTEs/subqueries.  It's easier to troubleshoot and allows you to break down a complex query.  

It also has the added benefit of being able to DROP temporary tables if you're working with very large CSV files as everything is done in-memory by `sqlp`.  Though it can do streaming operations, its always better to have plenty of memory headroom. 

```SQL
-- first, create a table from the select
CREATE TABLE DomainsExclude_AppStatus as 
SELECT DomainsExclude.name, DomainsExclude.appstatus
 FROM DomainsExclude, DomainsInclude
 WHERE DomainsInclude.name = DomainsExclude.name
 ORDER BY DomainsExclude.name;

-- then we do the join with the just created table
CREATE TABLE ResultTable as 
 SELECT *
  FROM DomainsInclude JOIN DomainsExclude_AppStatus
  ON DomainsInclude.name = DomainsExclude_AppStatus.name;

-- drop DomainsExclude_AppStatus
-- all the tables created during the script are automatically dropped at the end
-- but you can manually drop tables inside the script to free up memory
DROP TABLE DomainsExclude_AppStatus;

-- Group and sort the results
SELECT category, count(*) as category_cnt, *
 FROM ResultTable
 GROUP BY category
 ORDER BY category_cnt desc, category asc;
```

**Reply by ondohotola** [reply]
something like

```bash
qsv joinp -Q --left-semi 'Client ID' Clients.csv registrar_id DomainsInclude.csv
```

as the column has a space in ot I am struggling with SQL, but this is clean and quick

**Reply by derekmahar** [reply]
You can refer to a column name that contains a space by enclosing the column name in double quotes:

```
$ cat example_1.csv
Column 1,Column 2
a,1
b,2
c,3
$ qsv sqlp --quiet example_1.csv 'SELECT "Column 2", "Column 1" FROM example_1'
Column 2,Column 1
1,a
2,b
3,c
```
